velox/functions/sparksql/String.h

```diff
@@ -1233,4 +1233,167 @@ struct SoundexFunction {
+      const arg_type<Varchar>& left,
+      const arg_type<Varchar>& right,
+      int32_t threshold) {
+    auto leftCodePoints = reinterpret_cast<const uint8_t*>(left.data());
+    auto rightCodePoints = reinterpret_cast<const uint8_t*>(right.data());
+    doCall<uint8_t>(
+        result,
+        leftCodePoints,
+        rightCodePoints,
+        left.size(),
+        right.size(),
+        threshold);
+  }
+
+  void call(
+      out_type<int32_t>& result,
+      const arg_type<Varchar>& left,
+      const arg_type<Varchar>& right) {
+    call(result, left, right, INT32_MAX);
+  }
+
+  void callAscii(
+      out_type<int32_t>& result,
+      const arg_type<Varchar>& left,
+      const arg_type<Varchar>& right) {
+    callAscii(result, left, right, INT32_MAX);
+  }
+
+ private:
+  // This implementation only computes the distance if it's less than or equal
+  // to the threshold value, setting result as -1 if it's greater.
+  // Threshold k allows us to reduce it to O(km) time by only computing a
+  // diagonal stripe of at most width 2k + 1 of the cost table.
+  // One example: suppose s is of length 5, t is of length 7,
+  // and our threshold is 1. In this case we're going to walk a stripe of
+  // length 3. The matrix would look like so:
+  //
+  // <pre>
+  //    0 1 2 3 4
+  // 0 |#|#| | | | --> lower boundary:0, upper:2;
+  // 1 |#|#|#| | | --> lower boundary:0, upper:3;
+  // 2 | |#|#|#| | --> lower boundary:1, upper:4;
+  // 3 | | |#|#|#| --> lower boundary:2, upper:5;
+  // 4 | | | |#|#| --> lower boundary:3, upper:5;
+  // 5 | | | | |#| --> lower boundary:4, upper:5;
+  // 6 | | | | | |
+  // </pre>
+  template <typename TCodePoint>
+  void doCall(
+      out_type<int32_t>& result,
+      const TCodePoint* leftCodePoints,
+      const TCodePoint* rightCodePoints,
+      size_t leftCodePointsSize,
+      size_t rightCodePointsSize,
+      int32_t threshold) {
+    if (leftCodePointsSize < rightCodePointsSize) {
+      doCall(
+          result,
+          rightCodePoints,
+          leftCodePoints,
+          rightCodePointsSize,
+          leftCodePointsSize,
+          threshold);
+      return;
+    }
+    VELOX_USER_CHECK_LE(
+        leftCodePointsSize,
+        INT32_MAX,
+        "The inputs size exceeded max Levenshtein distance input size,"
+        " the code points size of left is {}, code points size of right is {}",
+        leftCodePointsSize,
+        rightCodePointsSize);
+    if (leftCodePointsSize - rightCodePointsSize > threshold) {
+      result = -1;
+      return;
+    }
+    if (rightCodePointsSize == 0) {
+      result = leftCodePointsSize;
+      return;
+    }
+    std::vector<int32_t> distances;
+    distances.reserve(rightCodePointsSize);
+    // These fills ensure that the value above the rightmost entry of our
+    // stripe will be ignored in following loop iterations.
+    int32_t boundary = std::min<int32_t>(rightCodePointsSize, threshold);
+    auto i = 0;
+    for (; i < boundary; i++) {
```

Comment:
If we remove the boundary variable, I think this may cause min function is evaluated multi times in the for loop. Please let mw know if this makes sense to you.
Thanks.

Reply:
Makes sense. Thanks.